core/chain/src/lib.rs

```diff
@@ -1,27 +1,64 @@
 extern crate parking_lot;
 extern crate primitives;
+extern crate serde;
 extern crate storage;
 
 use std::collections::HashMap;
+use std::fmt::Debug;
 use std::sync::Arc;
 
 use parking_lot::RwLock;
+use serde::{de::DeserializeOwned, Serialize};
 
 use primitives::hash::CryptoHash;
-use primitives::traits::{Block, Decode, Encode, Header};
-use primitives::types::BlockId;
+use primitives::traits::{Decode, Encode, Signer};
+use primitives::types::{BlockId, PartialSignature};
 use primitives::utils::index_to_bytes;
 use storage::Storage;
 
-const BLOCKCHAIN_GENESIS_BLOCK: &[u8] = b"genesis";
 const BLOCKCHAIN_BEST_BLOCK: &[u8] = b"best";
 
+/// Trait that abstracts ``Header"
+pub trait Header: Debug + Clone + Send + Sync + Serialize + DeserializeOwned + Eq + 'static
+{
+    /// Returns hash of the block body.
+    fn block_hash(&self) -> CryptoHash;
+
+    /// Returns block index.
+    fn index(&self) -> u64;
+
+    /// Returns hash of parent block.
+    fn parent_hash(&self) -> CryptoHash;
+}
+
+/// Trait that abstracts a ``Block", Is used for both beacon-chain blocks
+/// and shard-chain blocks.
+pub trait Block: Debug + Clone + Send + Sync + Serialize + DeserializeOwned + Eq + 'static {
```

Comment:
Why do we need it to be clonable? Blocks are pretty heavy so we should be careful with cloning them.

Reply:
We are clonning them at the moment. If we figure out how to not do that, e.g.:
```
fn write_with_cache<T: Clone + Encode>(
    storage: &Arc<Storage>,
    col: Option<u32>,
    cache: &RwLock<HashMap<Vec<u8>, T>>,
    key: &[u8],
    value: &T,
) {
    let data = Encode::encode(value).expect("Error serializing data");
    cache.write().insert(key.to_vec(), value.clone());

    let mut db_transaction = storage.transaction();
    db_transaction.put(col, key, &data);
    storage.write(db_transaction).expect("Database write failed");
}
```
we can remove it later.

Reply:
Filed https://github.com/nearprotocol/nearcore/issues/179